File: spark/src/main/scala/geotrellis/spark/io/file/FileRDDWriter.scala
package geotrellis.spark.io.file

import geotrellis.spark.io.avro.{AvroRecordCodec, AvroEncoder}
import geotrellis.spark.io.avro.codecs.KeyValueRecordCodec
import geotrellis.util.Filesystem

import org.apache.spark.rdd.RDD

import scala.reflect.ClassTag
import java.io.File

object FileRDDWriter {
  def write[K: AvroRecordCodec: ClassTag, V: AvroRecordCodec: ClassTag](
    rdd: RDD[(K, V)],
    rootPath: String,
    keyPath: K => String
  ): Unit = {
    val codec  = KeyValueRecordCodec[K, V]
    val schema = codec.schema

    val pathsToTiles =
      // Call groupBy with numPartitions; if called without that argument or a partitioner,
      // groupBy will reuse the partitioner on the parent RDD if it is set, which could be typed
      // on a key type that may no longer by valid for the key type of the resulting RDD.
      rdd.groupBy({ row => keyPath(row._1) }, numPartitions = rdd.partitions.length)

    Filesystem.ensureDirectory(rootPath)

    pathsToTiles.foreach { case (path, rows) =>
      val bytes = AvroEncoder.toBinary(rows.toVector)(codec)
      Filesystem.writeBytes(path, bytes)
    }
  }
}

File: geotrellis-python/geotrellis/spark/io/file/FileRDDWriter.py
from __future__ import absolute_import
from geotrellis.spark.io.avro.codecs.KeyValueRecordCodec import KeyValueRecordCodec
from geotrellis.util.Filesystem import Filesystem
from geotrellis.spark.io.avro.AvroEncoder import AvroEncoder

class FileRDDWriter(object):

    @staticmethod
    def write(K, V, rdd, rootPath, keyPath):
        codec = KeyValueRecordCodec(K, V)
        schema = codec.schema

        pathsToFiles = rdd.groupBy(
                lambda row: keyPath(row[0]),
                numPartitions = rdd.getNumPartitions())
        Filesystem.ensureDirectory(rootPath)

        def func(t):
            path, rows = t
            bytesArray = AvroEncoder.toBinary(rows, codec)
            Filesystem.writeBytes(path, bytesArray)

        pathsToFiles.foreach(func)
